syntax = "proto3";

package chubby;

service Chubby {
    rpc AcquireLock(AcquireLockRequest) returns (AcquireLockResponse);
    rpc TryAcquireLock(TryAcquireLockRequest) returns (TryAcquireLockResponse);
    rpc ReleaseLock(ReleaseLockRequest) returns (ReleaseLockResponse);
    rpc KeepAlive(KeepAliveRequest) returns (KeepAliveResponse);
}

enum LockingMode {
    SHARED = 0;
    EXCLUSIVE = 1;
}

message AcquireLockRequest {
    string client_id = 1;
    string path = 2;
    LockingMode locking_mode = 3; 
}

message AcquireLockResponse {
    bool success = 1;
    string error_message = 2;
}

message TryAcquireLockRequest {
    string path = 1;
}

message TryAcquireLockResponse {
    bool success = 1;
    string error_message = 2;
}

message ReleaseLockRequest {
    string path = 1;
}

message ReleaseLockResponse {
    bool success = 1;
    string error_message = 2;
}

message KeepAliveRequest {
    string client_id = 1;
}

message KeepAliveResponse {
    bool success = 1;              
    uint32 lease_timeout = 2;     
    string error_message = 3; 
}